account-compression/programs/spl-compression/src/lib.rs

```diff
@@ -0,0 +1,550 @@
+        match ($header.max_depth, $header.max_buffer_size) {
+            (3, 8) => merkle_roll_depth_size_apply_fn!(3, 8, $id, $bytes, $func, $($arg)*),
+            (5, 8) => merkle_roll_depth_size_apply_fn!(5, 8, $id, $bytes, $func, $($arg)*),
+            (14, 64) => merkle_roll_depth_size_apply_fn!(14, 64, $id, $bytes, $func, $($arg)*),
+            (14, 256) => merkle_roll_depth_size_apply_fn!(14, 256, $id, $bytes, $func, $($arg)*),
+            (14, 1024) => merkle_roll_depth_size_apply_fn!(14, 1024, $id, $bytes, $func, $($arg)*),
+            (14, 2048) => merkle_roll_depth_size_apply_fn!(14, 2048, $id, $bytes, $func, $($arg)*),
+            (20, 64) => merkle_roll_depth_size_apply_fn!(20, 64, $id, $bytes, $func, $($arg)*),
+            (20, 256) => merkle_roll_depth_size_apply_fn!(20, 256, $id, $bytes, $func, $($arg)*),
+            (20, 1024) => merkle_roll_depth_size_apply_fn!(20, 1024, $id, $bytes, $func, $($arg)*),
+            (20, 2048) => merkle_roll_depth_size_apply_fn!(20, 2048, $id, $bytes, $func, $($arg)*),
+            (24, 64) => merkle_roll_depth_size_apply_fn!(24, 64, $id, $bytes, $func, $($arg)*),
+            (24, 256) => merkle_roll_depth_size_apply_fn!(24, 256, $id, $bytes, $func, $($arg)*),
+            (24, 512) => merkle_roll_depth_size_apply_fn!(24, 512, $id, $bytes, $func, $($arg)*),
+            (24, 1024) => merkle_roll_depth_size_apply_fn!(24, 1024, $id, $bytes, $func, $($arg)*),
+            (24, 2048) => merkle_roll_depth_size_apply_fn!(24, 2048, $id, $bytes, $func, $($arg)*),
+            (26, 512) => merkle_roll_depth_size_apply_fn!(26, 512, $id, $bytes, $func, $($arg)*),
+            (26, 1024) => merkle_roll_depth_size_apply_fn!(26, 1024, $id, $bytes, $func, $($arg)*),
+            (26, 2048) => merkle_roll_depth_size_apply_fn!(26, 2048, $id, $bytes, $func, $($arg)*),
+            (30, 512) => merkle_roll_depth_size_apply_fn!(30, 512, $id, $bytes, $func, $($arg)*),
+            (30, 1024) => merkle_roll_depth_size_apply_fn!(30, 1024, $id, $bytes, $func, $($arg)*),
+            (30, 2048) => merkle_roll_depth_size_apply_fn!(30, 2048, $id, $bytes, $func, $($arg)*),
+            _ => {
+                msg!("Failed to apply {} on merkle roll with max depth {} and max buffer size {}", stringify!($func), $header.max_depth, $header.max_buffer_size);
+                err!(AccountCompressionError::ConcurrentMerkleTreeConstantsError)
+            }
+        }
+    };
+}
+
+#[program]
+pub mod spl_compression {
+    use super::*;
+
+    /// Creates a new merkle tree with maximum leaf capacity of `power(2, max_depth)`
+    /// and a minimum concurrency limit of `max_buffer_size`.
+    ///
+    /// Concurrency limit represents the # of replace instructions that can be successfully
+    /// executed with proofs dated for the same root. For example, a maximum buffer size of 1024
+    /// means that a minimum of 1024 replaces can be executed before a new proof must be
```

Comment:
Actually... this is a minimum.

SPL CMTs by default will fast-forward proofs through the entire changelog buffer if it can't find the original root that proof was submitted for. This can create scenarios where proofs can be more then BUFFER_SIZE (1024) operations old, yet still be fast forwarded to a valid proof...

This feels a little bit unsafe, but the operating thesis was that so long as you know what the leaf data is, you should be able to modify it.

Reply:
Does that mean you literally *can't* generate another proof before 1024 replaces?

Reply:
I think "at least" is the better term. 

It's 1024 replaces before "cache entry" is flushed. However based on access pattern of nearby leaves, we may be able to recreate the cache entry as needed.